Search json output: conan search --json output.json

danimtb opened the pull request:
- [x] Refer to the issue that supports this Pull Request: closes #2791
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

Tasks performed:

- Raise error if searching in local cache for packages and recipe is not found. This was behaving in a different way than searching packages in a remote. Now both cases raise error "Recipe not found".
- Recipe hashes are now calculated always, no matter local cache/remote or recipe/packages search.
- Added try/except for error in conan API and search command to manage output well with exceptions.
- Modified return od search functions, now returning info for the output (this makes the return of conan API search methos more homogeneous).
- Added `--json` parameter to `conan search`.

### Output examples
- Search references in all remotes:
`$ conan search eigen* -r all`

```
{  
    "error":false,
    "results":[  
        {  
            "remote":"conan-center",
            "items":[  
                {  
                    "recipe":{  
                        "id":"eigen/3.3.4@conan/stable"
                    }
                }
            ]
        },
        {  
            "remote":"upload_repo",
            "items":[  
                {  
                    "recipe":{  
                        "id":"eigen/3.3.4@danimtb/stable"
                    }
                },
                {  
                    "recipe":{  
                        "id":"eigen/3.3.4@danimtb/testing"
                    }
                }
            ]
        },
        {  
            "remote":"conan-community",
            "items":[  
                {  
                    "recipe":{  
                        "id":"eigen/3.3.4@conan/stable"
                    }
                }
            ]
        }
    ]
}
```
- Search packages of a reference in a remote:
`$ conan search paho-c/1.2.0@conan/stable -r conan-center --json search.json`

```
{
    "error":false,
    "results":[
        {
            "remote":"conan-center",
            "items":[
                {
                    "recipe":{
                        "id":"paho-c/1.2.0@conan/stable"
                    },
                    "packages":[
                        {
                            "id":"0000193ac313953e78a4f8e82528100030ca70ee",
                            "options":{
                                "shared":"False",
                                "asynchronous":"False",
                                "SSL":"False"
                            },
                            "settings":{
                                "os":"Linux",
                                "arch":"x86_64",
                                "compiler":"gcc",
                                "build_type":"Debug",
                                "compiler.version":"4.9"
                            },
                            "requires":[

                            ],
                            "outdated":false
                        },
                        {
                            "id":"014be746b283391f79d11e4e8af3154344b58223",
                            "options":{
                                "shared":"False",
                                "asynchronous":"False",
                                "SSL":"False"
                            },
                            "settings":{
                                "os":"Windows",
                                "compiler.threads":"posix",
                                "compiler.exception":"seh",
                                "arch":"x86_64",
                                "compiler":"gcc",
                                "build_type":"Debug",
                                "compiler.version":"5"
                            },
                            "requires":[

                            ],
                            "outdated":false
                        },
                        {
                            "id":"0188020dbfd167611b967ad2fa0e30710d23e920",
                            "options":{
                                "shared":"True",
                                "asynchronous":"False",
                                "SSL":"False"
                            },
                            "settings":{
                                "os":"Macos",
                                "arch":"x86_64",
                                "compiler":"apple-clang",
                                "build_type":"Debug",
                                "compiler.version":"9.1"
                            },
                            "requires":[

                            ],
                            "outdated":false
                        },
                        {
                            "id":"03369b0caf8c0c8d4bb84d5136112596bde4652d",
                            "options":{
                                "shared":"True",
                                "asynchronous":"False",
                                "SSL":"False"
                            },
                            "settings":{
                                "os":"Linux",
                                "arch":"x86",
                                "compiler":"gcc",
                                "build_type":"Release",
                                "compiler.version":"5"
                            },
                            "requires":[

                            ],
                            "outdated":false
                        }
                    ]
                }
            ]
        }
    ]
}
```
- Search references in local cache:
`conan search paho-c* --json search.json`

```
{
    "error":false,
    "results":[
        {
            "remote":"None",
            "items":[
                {
                    "recipe":{
                        "id":"paho-c/1.2.0@danimtb/testing"
                    }
                }
            ]
        }
    ]
}
```

- Search packages of a reference in local cache:
`$ conan search paho-c/1.2.0@danimtb/testing --json search.json`

```
{
    "error":false,
    "results":[
        {
            "remote":"None",
            "items":[
                {
                    "recipe":{
                        "id":"paho-c/1.2.0@danimtb/testing"
                    },
                    "packages":[
                        {
                            "id":"6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7",
                            "options":{
                                "SSL":"False",
                                "asynchronous":"False",
                                "shared":"False"
                            },
                            "settings":{
                                "arch":"x86_64",
                                "build_type":"Release",
                                "compiler":"Visual Studio",
                                "compiler.runtime":"MD",
                                "compiler.version":"15",
                                "os":"Windows"
                            },
                            "requires":[

                            ],
                            "outdated":false
                        },
                        {
                            "id":"95cd13dfc3f6b80d3ccb2a38441e3a1ad88e5a15",
                            "options":{
                                "SSL":"False",
                                "asynchronous":"True",
                                "shared":"True"
                            },
                            "settings":{
                                "arch":"x86_64",
                                "build_type":"Release",
                                "compiler":"Visual Studio",
                                "compiler.runtime":"MD",
                                "compiler.version":"15",
                                "os":"Windows"
                            },
                            "requires":[

                            ],
                            "outdated":true
                        },
                        {
                            "id":"970e773c5651dc2560f86200a4ea56c23f568ff9",
                            "options":{
                                "SSL":"False",
                                "asynchronous":"False",
                                "shared":"True"
                            },
                            "settings":{
                                "arch":"x86_64",
                                "build_type":"Release",
                                "compiler":"Visual Studio",
                                "compiler.runtime":"MD",
                                "compiler.version":"15",
                                "os":"Windows"
                            },
                            "requires":[

                            ],
                            "outdated":true
                        },
                        {
                            "id":"c4c0a49b09575515ce1dd9841a48de0c508b9d7c",
                            "options":{
                                "SSL":"True",
                                "asynchronous":"False",
                                "shared":"True"
                            },
                            "settings":{
                                "arch":"x86_64",
                                "build_type":"Release",
                                "compiler":"Visual Studio",
                                "compiler.runtime":"MD",
                                "compiler.version":"15",
                                "os":"Windows"
                            },
                            "requires":[
                                "OpenSSL/1.0.2n@conan/stable:606fdb601e335c2001bdf31d478826b644747077",
                                "zlib/1.2.11@conan/stable:6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7"
                            ],
                            "outdated":true
                        },
                        {
                            "id":"db9d6ba7004592ed2598f2c369484d4a01269110",
                            "options":{
                                "SSL":"True",
                                "asynchronous":"False",
                                "shared":"True"
                            },
                            "settings":{
                                "arch":"x86_64",
                                "build_type":"Release",
                                "compiler":"gcc",
                                "compiler.exception":"seh",
                                "compiler.threads":"posix",
                                "compiler.version":"7",
                                "os":"Windows"
                            },
                            "requires":[
                                "OpenSSL/1.0.2n@conan/stable:f761d91cef7988eafb88c6b6179f4cf261609f26",
                                "zlib/1.2.11@conan/stable:6dc82da13f94df549e60f9c1ce4c5d11285a4dff"
                            ],
                            "outdated":true
                        }
                    ]
                }
            ]
        }
    ]
}
```

### Comments and Concerns
- I think showing the recipe hash is useful to identify same recipes in different remotes, for example, the `eigen/3.3.4@conan/stable` is in conan-center remote and in conan-community
- Option values are strings (`"True"`) instead of just booleans (`True`).
- When searching in local cache it could be confussing to show a first level `remote : None`. However, this was made to maintain a more general output for all search optionsin command line.

**EDIT:** Removed recipe hashes. Updated json output

memsharded commented on conans/client/recorder/search_recorder.py:
```diff
@@ -0,0 +1,58 @@
+                "requires": self.requires, "outdated": self.outdated}
+
+
+class SearchRecoder(object):
```
Recoder => Recorder

memsharded commented on conans/client/cmd/search.py:
```diff
@@ -11,23 +11,36 @@ def __init__(self, client_cache, remote_manager, remote_registry):
+        for ref in refs:
+            recipe_hash = self._get_recipe_hash_remote(ref, remote)
+            references[remote.name].append((ref, recipe_hash))
+        return OrderedDict(references)
```
It doesn't make sense, sometimes you return an OrderedDict, other you return a dict.

Building an OrderedDict from a dict makes no sense (regarding ordering)

memsharded commented on conans/client/cmd/search.py:
```diff
@@ -11,23 +11,36 @@ def __init__(self, client_cache, remote_manager, remote_registry):
+                    if refs:
+                        references[remote.name] = []
+                        for ref in refs:
+                            recipe_hash = self._get_recipe_hash_remote(ref, remote)
```
This would be terribly slow for recipe-search with multiple results

memsharded commented on conans/client/command.py:
```diff
@@ -821,31 +822,53 @@ def search(self, *args):
+                self._outputer.print_search_packages(info["results"], reference, args.query,
+                                                     args.table)
+            else:
+                if args.table:
```
Maybe the --table and --json args are totally incompatible, shouldn't get a json captured exception but a command exception

memsharded commented on conans/client/conan_api.py:
```diff
@@ -668,14 +669,45 @@ def users_list(self, remote=None):
+                                                                        outdated=outdated)
+        except ConanException as exc:
+            recorder.error = True
+            exc.info = recorder.get_info()
```
Won't get_info() be empty here? What is this for? Doesn't make sense.

danimtb replied:
Yes, but you have to return the error to the excetion output to indicate in the json that the command has failed. You cant try to add information to the recorder as you dont have the return objects of the `search.search_recipe()` call

memsharded commented on conans/client/recorder/search_recorder.py:
```diff
@@ -0,0 +1,58 @@
+class _SearchRecipe(namedtuple("SearchRecipe", "reference, recipe_hash")):
+    with_packages = True
+
+    def __new__(cls, reference, recipe_hash):
```
Why is this new necessary? Shouldn't be

memsharded commented on conans/search/search.py:
```diff
@@ -103,6 +103,8 @@ def search_packages(paths, reference, query):
                            settings: {os: Windows}}}
     param conan_ref: ConanFileReference object
     """
+    if not search_recipes(paths, reference):
```
Dangerous, this search by pattern, why is this necessary?

danimtb replied:
This is done to keep consistency with a remote `search_packages()`. If you search packages of a recipe that does not exist in the remote, the current behavior raises "Recipe not found". This was not happening with a local cache `search_packages()`

lasote replied:
And what about check directly if the recipe folder is present?

danimtb commented:
- Removed all hashes from json output.
- Removed OrderedDict()
- Fix Recorder class name
- Raise exception for incompatible arguments `--json` and `--table`

lasote commented on conans/client/cmd/search.py:
```diff
@@ -11,23 +11,31 @@ def __init__(self, client_cache, remote_manager, remote_registry):
-            return search_recipes(self._client_cache, pattern, ignorecase)
+            references[None] = []
+            refs = search_recipes(self._client_cache, pattern, ignorecase)
+            [references[None].append(ref) for ref in refs]
```
`references[None].extend(refs)` or am I missing something?

lasote commented on conans/client/cmd/search.py:
```diff
@@ -11,23 +11,31 @@ def __init__(self, client_cache, remote_manager, remote_registry):
-        return self._remote_manager.search_recipes(remote, pattern, ignorecase)
+        references[remote.name] = []
+        refs = self._remote_manager.search_recipes(remote, pattern, ignorecase)
+        [references[remote.name].append(ref) for ref in refs]
```
Same here

lasote commented on conans/client/cmd/search.py:
```diff
@@ -11,23 +11,31 @@ def __init__(self, client_cache, remote_manager, remote_registry):
+                    refs = self._remote_manager.search_recipes(remote, pattern, ignorecase)
+                    if refs:
+                        references[remote.name] = []
+                        [references[remote.name].append(ref) for ref in refs]
```
same here
